File: Serie2.java/src/main/java/ThrottledRegion.java
import java.util.LinkedList;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class ThrottledRegion {

    private static class Item {
        public final Condition condition;
        public boolean done = false;

        public Item(Condition condition) {
            this.condition = condition;
        }
    }

    private class ThrottledRegionForKey {

        private final AtomicInteger _insideCount = new AtomicInteger(0);

        // changed inside the lock but observed outside of it
        private volatile int _waitingCount = 0;
        private final Lock _lock = new ReentrantLock();

        // changed and observed inside the lock
        private final LinkedList<Item> _waitingQueue = new LinkedList<>();


        /*
        |--------------------------------------------------------------------------
        | Try Enter
        |--------------------------------------------------------------------------
        */
        public boolean tryEnter() throws InterruptedException {
            // fast path
            if (_waitingCount == 0 && tryAcquireInside()) {
                return true;
            }

            // slow path
            _lock.lock();
            try {
                // if full, quit immediately
                if (_waitingCount == _maxWaiting) {
                    return false;
                }

                // mark as waiting
                // until this point, this thread can be overtaken by another one in the fast path
                _waitingCount += 1;

                // If first in waiting line, check again if region is not full
                if (_waitingCount == 1 && tryAcquireInside()) {
                    _waitingCount -= 1;
                    return true;
                }

                // waiting..
                Item item = new Item(_lock.newCondition());
                _waitingQueue.addLast(item);
                long timeout = _waitTimeout;
                do {
                    try {
                        timeout = item.condition.awaitNanos(timeout);
                    } catch (InterruptedException e) {
                        if (item.done) {
                            Thread.currentThread().interrupt();
                            return true;
                        }

                        _waitingCount -= 1;
                        _waitingQueue.remove(item);
                        throw e;
                    }

                    if (item.done) {
                        return true;
                    }

                    if (timeout <= 0) {
                        _waitingCount -= 1;
                        _waitingQueue.remove(item);
                        return false;
                    }
                } while (true);
            } finally {
                _lock.unlock();
            }
        }

        /*
        |--------------------------------------------------------------------------
        | Leave
        |--------------------------------------------------------------------------
        */
        public void leave() {
            boolean alreadyDecremented = false;
            if (_waitingCount == 0) {
                _insideCount.decrementAndGet();
                if (_waitingCount == 0) {
                    return;
                }

                alreadyDecremented = true;
            }

            _lock.lock();
            try {
                // check again if no one is waiting
                Item first = _waitingQueue.peekFirst();
                if (first == null) {
                    _insideCount.decrementAndGet();
                    return;
                }

                // someone is waiting but no space
                if (alreadyDecremented && !tryAcquireInside()) {
                    return;
                }

                first.done = true;
                _waitingCount -= 1;
                _waitingQueue.remove(first);
                first.condition.signal();
            } finally {
                _lock.unlock();
            }
        }

        private boolean tryAcquireInside() {
            do {
                int observed = _insideCount.get();
                if (observed >= _maxInside) {
                    return false;
                }
                if (_insideCount.compareAndSet(observed, observed + 1)) {
                    return true;
                }
            } while (true);
        }

    }

    private final int _maxInside;
    private final int _maxWaiting;
    private final int _waitTimeout;
    private final ConcurrentMap<Integer, ThrottledRegionForKey> _keyToRegion = new ConcurrentHashMap<>();

    public ThrottledRegion(int maxInside, int maxWaiting, int waitTimeout) {
        this._maxInside = maxInside;
        this._maxWaiting = maxWaiting;
        this._waitTimeout = waitTimeout; // convert millis to seconds?
    }

    public boolean tryEnter(int key) throws InterruptedException {
        return _keyToRegion.computeIfAbsent(key, k -> new ThrottledRegionForKey()).tryEnter();
    }

    public void leave(int key) {
        _keyToRegion.get(key).leave();
    }
}

File: Serie2/Serie2/ThrottledRegion.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Serie2 {

    /*
    |--------------------------------------------------------------------------
    | Region for each key
    |--------------------------------------------------------------------------
    */

    internal class ThrottledRegionForKey {

        private readonly int _maxWaiting;
        private readonly int _waitTimeout;
        private readonly int _maxInside;

        // observed and changed outside the lock
        private volatile int _insideCount = 0;

        // changed inside the lock but observed outside of it
        private volatile int _waitingCount = 0;
        private readonly object _lock = new object();

        // changed and observed inside the lock
        private readonly LinkedList<bool> _waitingQueue = new LinkedList<bool>();

        public ThrottledRegionForKey(int maxInside, int maxWaiting, int waitTimeout) {
            this._maxInside = maxInside;
            this._maxWaiting = maxWaiting;
            this._waitTimeout = waitTimeout;
        }

        /*
        |--------------------------------------------------------------------------
        | TryEnter
        |--------------------------------------------------------------------------
        */

        public bool TryEnter() {
            // fast path
            if (_waitingCount == 0 && tryAcquireInside()) {
                return true;
            }

            // slow path
            lock (_lock) {
                // if full, quit immediately
                if (_waitingCount == _maxWaiting) {
                    return false;
                }

                // mark as waiting
                // until this point, this thread can be overtaken by another one in the fast path
                _waitingCount += 1;

                // If first in waiting line, check again if region is not full
                if (_waitingCount == 1 && tryAcquireInside()) {
                    _waitingCount -= 1;
                    return true;
                }

                // waiting..
                LinkedListNode<bool> node = _waitingQueue.AddLast(false);
                int timeout = _waitTimeout;
                int lastTime = (timeout != Timeout.Infinite) ? Environment.TickCount : 0;

                do {
                    try {
                        SyncUtils.Wait(_lock, node, timeout);
                    } catch (ThreadInterruptedException e) {
                        _waitingCount -= 1;
                        _waitingQueue.Remove(node);

                        if (node.Value) {
                            Thread.CurrentThread.Interrupt();
                            return true;
                        }

                        throw;
                    }

                    if (node.Value) {
                        return true;
                    }

                    if (SyncUtils.AdjustTimeout(ref lastTime, ref timeout) == 0) {
                        _waitingCount -= 1;
                        _waitingQueue.Remove(node);
                        return false;
                    }
                } while (true);
            }
        }

        /*
        |--------------------------------------------------------------------------
        | Leave
        |--------------------------------------------------------------------------
        */

        public void leave() {
            bool alreadyDecremented = false;
            if (_waitingCount == 0) {
                Interlocked.Decrement(ref _insideCount);
                if (_waitingCount == 0) {
                    return;
                }

                alreadyDecremented = true;
            }

            lock (_lock) {
                // check again if no one is waiting
                LinkedListNode<bool> first = _waitingQueue.First;
                if (first == null) {
                    _insideCount--;
                    return;
                }

                // someone is waiting but no space
                if (alreadyDecremented && !tryAcquireInside()) {
                    return;
                }

                first.Value = true;
                _waitingCount -= 1;
                _waitingQueue.Remove(first);
                SyncUtils.Notify(_lock, first);
            }
        }

        private bool tryAcquireInside() {
            do {
                int observed = _insideCount;
                if (observed >= _maxInside) {
                    return false;
                }

                if (Interlocked.CompareExchange(ref _insideCount, observed + 1, observed) == observed) {
                    return true;
                }
            } while (true);
        }

    }


    public class ThrottledRegion {

        private readonly int _maxInside;
        private readonly int _maxWaiting;
        private readonly int _waitTimeout;
        private readonly ConcurrentDictionary<int, ThrottledRegionForKey> _keyToRegion;

        public ThrottledRegion(int maxInside, int maxWaiting, int waitTimeout) {
            this._maxInside = maxInside;
            this._maxWaiting = maxWaiting;
            this._waitTimeout = waitTimeout;
            this._keyToRegion = new ConcurrentDictionary<int, ThrottledRegionForKey>();
        }

        public bool TryEnter(int key) {
            var region = _keyToRegion.GetOrAdd(key,
                new ThrottledRegionForKey(this._maxInside, this._maxWaiting, this._waitTimeout));

            return region.TryEnter();
        }


        public void Leave(int key) {
            _keyToRegion[key].leave();
        }

    }

}